compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -669,7 +669,6 @@ fn layout_of_uncached<'tcx>(
                 let index = VariantIdx::new(0);
                 for field in &variants[index] {
                     assert!(!field.is_unsized());
-                    align = align.max(field.align);
 
                     // If all non-ZST fields have the same ABI, forward this ABI
                     if optimize && !field.is_zst() {
```

Comment:
This `if` ignores `field.is_zst()` but they could increase the alignment.
You need to check both orders, because right now it will accept an earlier field and still increase the alignment later.

The safe thing to do is actually to add a check *after* this entire loop, that resets the `Abi` to `Aggregate` if its "inherent `align`" (sadly idk if there's a method for that... you may be able to record the alignment of the only non-ZST field alongside `abi`?).

That way, you're not depending on stateful iteration and are instead "aggregating" everything together.